### PR TITLE
Added solvers as a list

### PR DESCRIPTION
**Pull request by timothyclarke**
Might address https://github.com/terraform-iaac/terraform-kubernetes-cert-manager/issues/5 without being quite so rigid

Moved from data.tf & yaml file to local so users can provide solvers as structured data

### CODE REVIEW

**Comment by timothyclarke** [comment]
@bohdantverdyi  Can you look at this one for allowing more control of the solvers. It gives a way of adding DNS01 per https://github.com/terraform-iaac/terraform-kubernetes-cert-manager/issues/5 without having to provide the full yaml. Someone just provide the solvers section. Possibly with `solvers = yamldecode(file("mysolvers.yml"))` 

Sorry if it looks a little odd. I'm regularly jumping between helm, terraform and terragrunt so the syntax can look a little strange at times

**Comment by bohdantverdyi** [comment]
I'll check this. In generally better idea to not have any files, only terraform syntax )

File: locals.tf [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    apiVersion = "cert-manager.io/v1"
+    kind = "ClusterIssuer"
+    metadata = {
+      name = var.cluster_issuer_name
+    }
+    spec = {
+      acme = {
+        # The ACME server URL
+        server = var.cluster_issuer_server
+        preferredChaini = "ISRG Root X1"
```
Comment (bohdantverdyi) [inline locals.tf:12]: Typo:

please change this `preferredChaini` -> `preferredChain`

File: variables.tf [hunk 2]
```diff
@@ -52,4 +52,68 @@ variable "cluster_issuer_yaml" {
 variable "additional_set" {
   description = "Additional sets to Helm"
   default     = []
-}
\ No newline at end of file
+}
+
+variable "solvers" {
+  description = "List of Cert manager solvers"
+  #type        = list(map(any))
```
Comment (bohdantverdyi) [inline variables.tf:59]: You can remove this. Terraform restriction to set up type needed for us

File: variables.tf [hunk 3]
```diff
@@ -52,4 +52,68 @@ variable "cluster_issuer_yaml" {
...
+  type        = any
+  default = [{
+    http01 = {
+      ingress = {
+        class = "nginx"
+      }
+    }
+  }]
+  #
```
Comment (bohdantverdyi) [inline variables.tf:68]: Good idea to add this into README.md. Like: how to work with multiple solvers.

Or you can remove it, i will update readme by myself

File: variables.tf [hunk 4]
```diff
@@ -52,4 +52,68 @@ variable "cluster_issuer_yaml" {
...
+  #   {
+  #     http01 = {
+  #       ingress = {
+  #         class = "nginx"
+  #       }
+  #     }
+  #   }
+  # ]
+  #
+  # FixMe  Define multiple object types where all are valid
```
Comment (bohdantverdyi) [inline variables.tf:110]: terraform restriction

Comment (bohdantverdyi) [reply variables.tf:110]: imposible to fix this

**Review by bohdantverdyi — CHANGES_REQUESTED** [review]
Everything works good. Few things which should be edited.

File: variables.tf (hunk 3 above)
Comment (bohdantverdyi) [reply variables.tf:67]:
```
# Example from https://github.com/terraform-iaac/terraform-kubernetes-cert-manager/issues/5
  #
  # [
  #   {
  #     dns01 = {
  #       route53 = {
  #         region  = "us-east-1"
  #         ambient = "true"
  #       }
  #     },
  #     selector = {
  #       dnsZones = [
  #         "internal.example.com"
  #       ]
  #     }
  #   },
  #   {
  #     dns01 = {
  #       cloudflare = {
  #         email = "user@example.com"
  #         apiKeySecretRef = {
  #           name = "cloudflare-api-key-secret"
  #           key  = "API"
  #         }
  #       },
  #     },
  #     selector = {
  #       dnsZones = [
  #         "public.example.com"
  #       ]
  #     }
  #   },
  #   {
  #     http01 = {
  #       ingress = {
  #         class = "nginx"
  #       }
  #     }
  #   }
  # ]
  #
```

I mean this part. We can delete this also.

Comment (bohdantverdyi) [reply variables.tf:67]:
```
# Example from https://github.com/terraform-iaac/terraform-kubernetes-cert-manager/issues/5
  #
  # [
  #   {
  #     dns01 = {
  #       route53 = {
  #         region  = "us-east-1"
  #         ambient = "true"
  #       }
  #     },
  #     selector = {
  #       dnsZones = [
  #         "internal.example.com"
  #       ]
  #     }
  #   },
  #   {
  #     dns01 = {
  #       cloudflare = {
  #         email = "user@example.com"
  #         apiKeySecretRef = {
  #           name = "cloudflare-api-key-secret"
  #           key  = "API"
  #         }
  #       },
  #     },
  #     selector = {
  #       dnsZones = [
  #         "public.example.com"
  #       ]
  #     }
  #   },
  #   {
  #     http01 = {
  #       ingress = {
  #         class = "nginx"
  #       }
  #     }
  #   }
  # ]
  #
```

I mean this part. We can delete this also.

**Comment by timothyclarke** [comment]
> Also, cloudflare supports two method of authentication, it's API Keys & API Token.
> 
> Thus if we want implement support of CloudFlare. There is a lot of logic required. We can do this, but do we want this ?

I think that way would be madness. As soon as you add CloudFlare people will be after other methods. Before long this one is going to be quite complex to show / support all the various combinations and permutations.

Adding the example you give above to the examples folder. Personally if there is enough call to extend to the cloudflare apikey/token then create a `terraform-kubernetes-cert-manager-cloudflare-apikey` repo that extends in that direction and consumes this repo. 
In the module I need to have a more complex `locals` to generate the selectors for different environments so I'd end up something more akin to
```
module "cert_manager" {
  source  = "terraform-iaac/cert-manager/kubernetes"
  solvers = local.solvers
```

**Comment by bohdantverdyi** [comment]
@timothyclarke Thank you for contributing. I appreciate this.

**Comment by timothyclarke** [comment]
@bohdantverdyi  Np, Easier to send you a PR than re-invent the wheel.
It would be great if we can get a tag on this so I can pull it in `source  = "terraform-iaac/cert-manager/kubernetes?ref=2.2.3" ;)

**Comment by bohdantverdyi** [comment]
> @bohdantverdyi  Np, Easier to send you a PR than re-invent the wheel.
> It would be great if we can get a tag on this so I can pull it in `source  = "terraform-iaac/cert-manager/kubernetes?ref=2.2.3"` ;)

new tag has been created: V2.3.0

**Pull request merged**